pkg/operator/ceph/object/bucket/label_controller.go

```diff
@@ -0,0 +1,271 @@
+			if err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: name}, bucketNotification); err != nil {
+				return errors.Wrapf(err, "failed to retrieve CephBucketNotification %q", name)
+			}
+			logger.Debugf("CephBucketNotification: %q resource found", bucketNotification.Name)
+
+			// Step 3: verify that the associated topic configuration exists with an ARN
+			bucketTopic := &cephv1.CephBucketTopic{}
+			if err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: bucketNotification.Spec.Topic}, bucketTopic); err != nil {
+				return errors.Wrapf(err, "failed to retrieve CephBucketTopic %q", name)
+			}
+			topicARN := topic.GetTopicARN(bucketTopic)
+			if topicARN == "" {
+				return errors.New("no ARN in topic. topic was not provisioned successfully")
+			}
+			parsedTopicARN, err := arn.Parse(topicARN)
+			if err != nil {
+				return err
+			}
+			if strings.ToLower(parsedTopicARN.Service) != "sns" {
+				return errors.New("topic ARN must have 'sns' service")
+			}
+			if parsedTopicARN.Resource == "" {
+				return errors.New("missing topic inside ARN")
+			}
+			logger.Debugf("CephBucketTopic found with valid ARN: %q", topicARN)
+
+			// step 4: get the ObjectBucket
+			cephOB := bktv1alpha1.ObjectBucket{}
+			if err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: cephOBC.Spec.ObjectBucketName}, &cephOB); err != nil {
+				return errors.Wrapf(err, "failed to retrieve ObjectBucket %q", cephOBC.Spec.ObjectBucketName)
+			}
+
+			// step 5: provision the notification
+			p := notification.Provisioner{
+				Client:      r.client,
+				Context:     r.context,
+				ClusterInfo: r.clusterInfo,
+				ClusterSpec: r.clusterSpec,
+			}
+			err = p.Create(&cephOB, topicARN, bucketNotification)
+			if err != nil {
+				return errors.Wrapf(err, "failed to provision bucket notification: %q", bucketNotification.Name)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (r *ReconcileOBCLabels) setFailedStatus(name types.NamespacedName, errMessage string, err error) (reconcile.Result, error) {
+	updateStatus(r.client, name, k8sutil.ReconcileFailedStatus)
+	return reconcile.Result{}, errors.Wrapf(err, "%s", errMessage)
+}
+
+// updateStatus updates the OBC with a given status
+func updateStatus(client client.Client, name types.NamespacedName, status bktv1alpha1.ObjectBucketClaimStatusPhase) {
+	cephOBC := &bktv1alpha1.ObjectBucketClaim{}
+	if err := client.Get(context.TODO(), name, cephOBC); err != nil {
+		if kerrors.IsNotFound(err) {
+			logger.Debug("ObjectBucketClaim resource not found. Ignoring since resource must be deleted.")
+			return
+		}
+		logger.Warningf("failed to retrieve ObjectBucketClaim %q to update status to %q. %v", name, status, err)
+		return
+	}
+
+	cephOBC.Status.Phase = status
+	if err := reporting.UpdateStatus(client, cephOBC); err != nil {
+		logger.Errorf("failed to set ObjectBucketClaim %q status to %q. %v", name, status, err)
+		return
+	}
```

Comment:
what about the case where the notification provisioning is failing - I still want the retry mechanism to work.
if i don't set the bucket status, it won't be retried.

Reply:
I don't understand why that must be. We can't assume that the OBC controller won't overwrite the status in its own operation. This should be a BucketNotification controller operation and retry, which should work without setting any statuses, though setting the failed status on the BN might not be a bad idea.

Reply:
i don't mind not setting the status (I actually removed it here: efed2619e14deef95f920601a38844e25b201c9e)
just want to make sure that the controller execution will be re-queued in case of error even if the status is not set to "failed"